src/main/java/co/cask/gcp/bigquery/BigQuerySource.java

```
@@ -221,63 +159,128 @@ public void onRunFinish(boolean succeeded, BatchSourceContext context) {
   @Path("getSchema")
   public Schema getSchema(BigQuerySourceConfig request) throws Exception {
     try {
-      BigQuery bigquery;
-      BigQueryOptions.Builder bigqueryBuilder = BigQueryOptions.newBuilder();
-      String serviceAccountFilePath = request.getServiceAccountFilePath();
-      if (serviceAccountFilePath != null) {
-        bigqueryBuilder.setCredentials(loadServiceAccountCredentials(serviceAccountFilePath));
-      }
-      String project = request.getProject();
-      if (project == null) {
-        throw new Exception("Could not detect Google Cloud project id from the environment. " +
-                              "Please specify a project id.");
+      Table table = BigQueryUtils.getBigqueryTable(request.getServiceAccountFilePath(), request.getProject(),
+                                                   request.dataset, request.table);
+      if (table == null) {
+        // Table does not exist
+        throw new IllegalArgumentException(String.format("BigQuery table %s does not exist", request.table));
       }
-      bigqueryBuilder.setProjectId(project);
-      bigquery = bigqueryBuilder.build().getService();
 
-      TableId id = TableId.of(project, request.dataset, request.table);
-      Table table = bigquery.getTable(id);
       com.google.cloud.bigquery.Schema bgSchema = table.getDefinition().getSchema();
+      List<Schema.Field> fields = getSchemaFields(bgSchema);
+      return Schema.recordOf("output", fields);
+    } catch (Exception e) {
+      throw new Exception(e.getMessage(), e);
```

doesn't look like there's a reason to catch this at all, just let it propagate.